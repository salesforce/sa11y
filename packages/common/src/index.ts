/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export { A11yConfig, AxeResults, axeRuntimeExceptionMsgPrefix, axeVersion, getAxeRules, getViolations } from './axe';
export { WdioAssertFunction, WdioOptions } from './wdio';
export { errMsgHeader, ExceptionList } from './format';
export { log, useFilesToBeExempted, useCustomRules, processFiles } from './helpers';
