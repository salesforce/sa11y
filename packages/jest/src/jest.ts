/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { assertAccessible } from '@sa11y/assert';
import { A11yConfig, extended } from '@sa11y/preset-rules';
import { matcherHint } from 'jest-matcher-utils';

export const matcherHintMsg = `expected document to have no accessibility violations but found following issues: `;

// Type def for custom jest a11y matcher toBeAccessible
// Ref: https://jestjs.io/docs/en/expect.html#expectextendmatchers
declare global {
    // eslint-disable-next-line @typescript-eslint/no-namespace
    namespace jest {
        interface Matchers<R> {
            toBeAccessible(config?: A11yConfig): Promise<CustomMatcherResult>;
            toBeAccessibleWith(config: A11yConfig): Promise<CustomMatcherResult>;
        }
    }
}

/**
 * Jest expect matcher to check DOM for accessibility issues
 * @param receivedDom - DOM to be tested for accessibility
 * @param config - A11yConfig to be used to test for accessibility. Defaults to extended.
 */
export async function toBeAccessible(
    receivedDom: Document = document,
    config: A11yConfig = extended
): Promise<jest.CustomMatcherResult> {
    let isAccessible = true;
    let a11yViolations = '';

    try {
        await assertAccessible(receivedDom, config);
    } catch (e) {
        isAccessible = false;
        a11yViolations = e;
    }
    return {
        pass: isAccessible,
        // Display assertion for the report when a test fails
        message: (): string => matcherHint(`${toBeAccessible.name}: ${matcherHintMsg} \n\n ${a11yViolations}`),
    };
}

/**
 * Jest expect matcher to check DOM for accessibility issues with a `@sa11y/preset-rule`
 * @param receivedDom - DOM to be tested for accessibility
 * @param config - A11yConfig to be used to test for accessibility.
 */
export function toBeAccessibleWith(
    receivedDom: Document = document,
    config: A11yConfig
): Promise<jest.CustomMatcherResult> {
    return toBeAccessible(receivedDom, config);
}
