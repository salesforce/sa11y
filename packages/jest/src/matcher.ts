/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { matcherHint, printReceived } from 'jest-matcher-utils';
import { adaptA11yConfig } from './setup';
import { a11yCheckableContext, assertAccessible } from '@sa11y/assert';
import { A11yError, Options } from '@sa11y/format';
import { A11yConfig, extended } from '@sa11y/preset-rules';

// Type def for custom jest a11y matchers
// Ref: https://jestjs.io/docs/en/expect.html#expectextendmatchers
declare global {
    // eslint-disable-next-line @typescript-eslint/no-namespace
    namespace jest {
        interface Matchers<R> {
            toBeAccessible(config?: A11yConfig): Promise<CustomMatcherResult>;
        }
    }
}

const matcherHintMsg = `expected document to have no accessibility violations but found`;
const expectedMsg = `0 issues`;
const FormatOptions: Options = {
    a11yViolationIndicator: 'â­•',
    helpUrlIndicator: 'ðŸ”—',
    highlighter: printReceived,
};

/**
 * Jest expect matcher to check DOM for accessibility issues
 * @param received - DOM or HTML Element to be tested for accessibility. Defaults to current DOM.
 * @param config - A11yConfig to be used to test for accessibility. Defaults to extended.
 */
export async function toBeAccessible(
    received: a11yCheckableContext = document,
    config: A11yConfig = extended
): Promise<jest.CustomMatcherResult> {
    let isAccessible = true;
    let a11yError: A11yError;
    let receivedMsg = expectedMsg;

    // TODO (Improvement): Can we detect if this is invoked async and error if not ?
    try {
        await assertAccessible(received, adaptA11yConfig(config));
    } catch (e) {
        if (e instanceof A11yError) {
            a11yError = e;
            isAccessible = false;
            receivedMsg = `${a11yError.length} issues`;
        } else {
            throw e;
        }
    }

    return {
        pass: isAccessible,
        message: (): string =>
            matcherHint(
                `${toBeAccessible.name}: ${matcherHintMsg} ${receivedMsg}: \n\n ${a11yError.format(FormatOptions)}`,
                `${a11yError.length} issues`,
                expectedMsg
            ),
    };
}
