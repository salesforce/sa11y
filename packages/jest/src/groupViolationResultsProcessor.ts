/*
 * Copyright (c) 2023, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { AggregatedResult, AssertionResult, TestResult } from '@jest/test-result';
import { log } from '@sa11y/common';
import { A11yError } from '@sa11y/format';

type FailureDetail = {
    error?: A11yError;
};
interface FailureMatcherDetail {
    error?: {
        matcherResult?: {
            a11yError?: A11yError;
        };
    };
}

type ErrorElement = {
    html: string;
    selectors: string;
    hierarchy: string;
    any: string;
    all: string;
    none: string;
};

type A11yViolation = {
    id: string;
    description: string;
    helpUrl: string;
    wcagCriteria: string;
    summary: string;
    errorElements: ErrorElement[];
};

const formatSpacing = '\t'.repeat(8);
const formatForAxeMessage = `\n${formatSpacing}\t\t`;

const axeMessages = {
    toSolveAny: `${formatForAxeMessage}- More Info: To solve the problem, you need to fix at least (1) of the following:\n`,
    toSolveFirst: `${formatForAxeMessage}- More Info: To solve the problem, you need to fix one of the following:\n`,
    toSolveSecond: `${formatForAxeMessage}- And fix the following:\n`,
};
/**
 * Create a test failure html elements array grouped by rule violation
 */
function createA11yErrorElements(errorElements: ErrorElement[]) {
    const a11yErrorElements: string[] = [];
    errorElements.forEach((errorElement, index) => {
        let errorMessage = `${formatSpacing}(${index + 1})  - HTML element : ${errorElement.html
            .replace(/&lt;/g, '<')
            .replace(/&gt;/, '>')}
                        - CSS selector(s) : ${errorElement.selectors.replace(/&gt;/, '>')}
                        - HTML Tag Hierarchy : ${errorElement.hierarchy}`;

        if (errorElement.any && errorElement.any.length > 0) {
            errorMessage += `${axeMessages.toSolveAny}${errorElement.any}`;
            if (errorElement.all && errorElement.all.length > 0) {
                errorMessage += `${axeMessages.toSolveSecond}${errorElement.all}`;
            }
            if (errorElement.none && errorElement.none.length > 0) {
                errorMessage += `${axeMessages.toSolveSecond}${errorElement.none}`;
            }
        } else if (errorElement.all && errorElement.all.length > 0) {
            errorMessage += `${axeMessages.toSolveFirst}${errorElement.all}`;
            if (errorElement.none && errorElement.none.length > 0) {
                errorMessage += `${axeMessages.toSolveSecond}${errorElement.none}`;
            }
        } else {
            if (errorElement.none && errorElement.none.length > 0) {
                errorMessage += `${axeMessages.toSolveFirst}${errorElement.none}`;
            }
        }
        a11yErrorElements.push(errorMessage);
    });

    return a11yErrorElements.join('\n');
}

/**
 * Create a test failure violation error message grouped by rule violation
 */
function createA11yRuleViolation(a11yRule: A11yViolation, ruleIndex: number) {
    return `(${ruleIndex}) [${a11yRule.id}] ${a11yRule.description}
            * Error element(s) : ${a11yRule.errorElements.length}\n${createA11yErrorElements(a11yRule.errorElements)}
            * Help:
                • Help URL: ${a11yRule.helpUrl}
                • WCAG Criteria: ${a11yRule.wcagCriteria}`;
}

/**
 * Create a test processA11yDetailsAndMessages violation error message grouped by rule violation
 */
function processA11yDetailsAndMessages(error: A11yError, a11yFailureMessages: string[]) {
    const a11yRuleViolations: { [key: string]: A11yViolation } = {};
    let a11yRuleViolationsCount = 0;
    let a11yErrorElementsCount = 0;
    error.a11yResults.forEach((a11yResult) => {
        a11yErrorElementsCount++;
        if (!(a11yRuleViolations as never)[a11yResult.wcag]) {
            a11yRuleViolationsCount++;
            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
            a11yRuleViolations[a11yResult.wcag] = {
                id: a11yResult.id,
                description: a11yResult.description,
                helpUrl: a11yResult.helpUrl,
                wcagCriteria: a11yResult.wcag,
                summary: a11yResult.summary,
                errorElements: [],
            };
        }
        a11yRuleViolations[a11yResult.wcag].errorElements.push({
            html: a11yResult.html,
            selectors: a11yResult.selectors,
            hierarchy: a11yResult.ancestry,
            any: a11yResult.any,
            all: a11yResult.all,
            none: a11yResult.none,
        });
    });

    const a11yFailureMessage = `
    The test has failed the accessibility check. Accessibility Stacktrace/Issues:
    ${a11yErrorElementsCount} HTML elements have accessibility issue(s). ${a11yRuleViolationsCount} rules failed.

    ${Object.values(a11yRuleViolations)
        .map((a11yRuleViolation, index) => createA11yRuleViolation(a11yRuleViolation, index + 1))
        .join('\n')}


    For more info about automated accessibility testing: https://sfdc.co/a11y-test
    For tips on fixing accessibility bugs: https://sfdc.co/a11y
    For technical questions regarding Salesforce accessibility tools, contact our Sa11y team: http://sfdc.co/sa11y-users
    For guidance on accessibility related specifics, contact our A11y team: http://sfdc.co/tmp-a11y
    `;
    a11yFailureMessages.push(a11yFailureMessage);
}

/**
 * Convert any a11y errors from test failures into their own test suite, results
 */
function processA11yErrors(results: AggregatedResult, testSuite: TestResult, testResult: AssertionResult) {
    const a11yFailureDetails: FailureDetail[] = [];
    const a11yFailureMessages: string[] = [];
    let a11yErrorsExist = false;

    testResult.failureDetails.forEach((failure) => {
        let error = (failure as FailureDetail).error;
        // If using circus test runner https://github.com/facebook/jest/issues/11405#issuecomment-843549606
        // TODO (code cov): Add test data covering the case for circus test runner
        /* istanbul ignore next */
        if (error === undefined) error = failure as A11yError;
        if (error.name === A11yError.name) {
            a11yErrorsExist = true;
            a11yFailureDetails.push({ ...(failure as FailureDetail) } as FailureDetail);
            processA11yDetailsAndMessages(error, a11yFailureMessages);
        }
    });
    if (!a11yErrorsExist) {
        testSuite.numFailingTests -= 1;
        results.numFailedTests -= 1;
        if (testSuite.numFailingTests === 0) results.numFailedTestSuites -= 1;
    }
    testResult.failureDetails = [...a11yFailureDetails];
    testResult.failureMessages = [...a11yFailureMessages];
    testResult.status = a11yFailureDetails.length === 0 ? 'passed' : testResult.status;
}

function processA11yMatcherErrors(results: AggregatedResult, testSuite: TestResult, testResult: AssertionResult) {
    const a11yFailureMessages: string[] = [];

    testResult.failureDetails.forEach((failure) => {
        const error = (failure as FailureMatcherDetail)?.error?.matcherResult?.a11yError as A11yError;
        if (error !== undefined) {
            processA11yDetailsAndMessages(error, a11yFailureMessages);
            testResult.failureMessages = [...a11yFailureMessages];
        }
    });
}

/**
 * Custom results processor for a11y results grouping the violations. Only affects JSON results file output.
 * To be used with jest cli options --json --outputFile
 *  e.g. jest --json --outputFile jestResults.json --testResultsProcessor `node_modules/@sa11y/jest/dist/groupViolationResultsProcessor.js`
 * Ref: https://jestjs.io/docs/configuration#testresultsprocessor-string
 *  - Mapping of AggregatedResult to JSON format to https://github.com/facebook/jest/blob/master/packages/jest-test-result/src/formatTestResults.ts
 */
export function resultsProcessor(results: AggregatedResult): AggregatedResult {
    log(`Processing ${results.numTotalTests} tests ..`);
    results.testResults // suite results
        .filter((testSuite) => testSuite.numFailingTests > 0)
        .forEach((testSuite) => {
            testSuite.testResults // individual test results
                .filter((testResult) => testResult.status === 'failed')
                .forEach((testResult) => {
                    processA11yErrors(results, testSuite, testResult);
                });
        });

    return results;
}

export function resultsProcessorManualChecks(results: AggregatedResult): AggregatedResult {
    log(`Processing ${results.numTotalTests} tests ..`);
    results.testResults // suite results
        .filter((testSuite) => testSuite.numFailingTests > 0)
        .forEach((testSuite) => {
            testSuite.testResults // individual test results
                .filter((testResult) => testResult.status === 'failed')
                .forEach((testResult) => {
                    processA11yMatcherErrors(results, testSuite, testResult);
                });
        });

    return results;
}
// The processor must be a node module that exports a function
// Explicitly typing the exports for Node.js compatibility
const exportedFunctions = {
    resultsProcessor,
    resultsProcessorManualChecks,
};

module.exports = exportedFunctions;
