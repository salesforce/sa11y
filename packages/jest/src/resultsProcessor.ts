/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { AggregatedResult, AssertionResult, TestResult, addResult, createEmptyTestResult } from '@jest/test-result';
import { errMsgHeader, log } from '@sa11y/common';
import { A11yError, A11yResult, A11yResults } from '@sa11y/format';

type FailureDetail = {
    error?: A11yError;
};

// This is used by downstream CI workflows to separate file/classname from attached WCAG metadata
const classNameMetadataSeparator = '::';
// Map of test suite name to test results
const consolidatedErrors = new Map<string, AssertionResult[]>();

/**
 * Create a test failure result with a11y meta-data at forefront from given test failure
 * with a11y error.
 */
function createA11yTestResult(testResult: AssertionResult, a11yResult: A11yResult): AssertionResult {
    return {
        ...testResult,
        // TODO (refactor): extract formatting into its own function.
        //  - Can this satisfy Formatter interface?
        //  - Be part of format? (FileFormatter vs ConsoleFormatter)?
        fullName: `${a11yResult.description}:${a11yResult.selectors}`,
        failureMessages: [
            `${errMsgHeader}: ${a11yResult.description}
Help: ${a11yResult.helpUrl}
CSS Selectors: ${a11yResult.selectors}
HTML element: ${a11yResult.html}
Tests: "${testResult.fullName}"
Summary: ${a11yResult.summary}`, // TODO (refactor): replace with array of tests when consolidating? But we reach here after de-duping a11y results?
        ],
        // We don't need the error objects anymore as they have been processed
        failureDetails: [],
        // Add all test's having the same a11y issue
        ancestorTitles: [...new Set(testResult.ancestorTitles).add(testResult.fullName)],
    } as AssertionResult;
}

/**
 * Convert any a11y errors from test failures into their own test suite, results
 */
function processA11yErrors(testSuite: TestResult, testResult: AssertionResult) {
    const suiteName = testSuite.testFilePath;

    testResult.failureDetails.forEach((failure) => {
        let error = (failure as FailureDetail).error;
        // If using circus test runner https://github.com/facebook/jest/issues/11405#issuecomment-843549606
        // TODO (code cov): Add test data covering the case for circus test runner
        /* istanbul ignore next */
        if (error === undefined) error = failure as A11yError;
        if (error.name === A11yError.name) {
            // TODO (spike) : What happens if there are ever multiple failureDetails?
            //  Ideally there shouldn't be as test execution should be stopped on failure
            A11yResults.add(error.a11yResults, suiteName).forEach((a11yResult) => {
                const className = `${suiteName}${classNameMetadataSeparator}${a11yResult.wcag}`;
                // TODO (code cov): Fix - should be covered by existing tests
                /* istanbul ignore next */
                if (!Array.isArray(consolidatedErrors.get(className))) consolidatedErrors.set(className, []);
                consolidatedErrors.get(className)?.push(createA11yTestResult(testResult, a11yResult));
            });
        }
    });
}

/**
 * Modify existing test result containing a11y error after a11y error
 *  is extracted into its own test result using {@link createA11yTestResult}.
 */
function modifyOriginalTestResult(results: AggregatedResult, testSuite: TestResult, testResult: AssertionResult) {
    // Don't report the failure twice
    testResult.status = 'disabled';
    // TODO (fix): Remove sa11y msg from test suite message.
    //  ANSI codes and test names in suite message makes it difficult.
    //  Removing error from test result doesn't affect test suite error msg.
    // testResult.failureMessages = [];
    // testResult.failureDetails = [];
    // testSuite.failureMessage = '';
    // Suites with only a11y errors should be marked as passed
    testSuite.numFailingTests -= 1;
    results.numFailedTests -= 1;
    if (testSuite.numFailingTests === 0) results.numFailedTestSuites -= 1;
    // TODO(debug): Does 'success' represent only failed tests?
    //  Or errored tests as well e.g.?
    // if (results.numFailedTestSuites === 0) results.success = true;
}

/**
 * Custom results processor for a11y results. Only affects JSON results file output.
 * To be used with jest cli options --json --outputFile
 *  e.g. jest --json --outputFile jestResults.json --testResultsProcessor `node_modules/@sa11y/jest/dist/resultsProcessor.js`
 * Ref: https://jestjs.io/docs/configuration#testresultsprocessor-string
 *  - Mapping of AggregatedResult to JSON format to https://github.com/facebook/jest/blob/master/packages/jest-test-result/src/formatTestResults.ts
 */
export default function resultsProcessor(results: AggregatedResult): AggregatedResult {
    // TODO (refactor): Use map/filter to get results directly without global var for consolidated errors
    log(`Processing ${results.numTotalTests} tests ..`);
    results.testResults // suite results
        .filter((testSuite) => testSuite.numFailingTests > 0)
        .forEach((testSuite) => {
            testSuite.testResults // individual test results
                .filter((testResult) => testResult.status === 'failed')
                .forEach((testResult) => {
                    processA11yErrors(testSuite, testResult);
                    modifyOriginalTestResult(results, testSuite, testResult);
                });
        });

    log(`Transforming a11y failures from ${consolidatedErrors.size} suites ..`);
    // Create test suites to hold a11y failures
    consolidatedErrors.forEach((testResults, suiteKey) => {
        // TODO (refactor): Do we need to create a test suite if suite name
        //  is not changing? Can this be simplified by adding tests to existing suite?
        const sa11ySuite = createEmptyTestResult();
        // "testFilePath" gets output as suite "name" in formatted JSON result
        sa11ySuite.testFilePath = suiteKey;
        sa11ySuite.testResults = testResults;
        sa11ySuite.numFailingTests = testResults.length;
        addResult(results, sa11ySuite);
    });

    return results;
}

// The processor must be a node module that exports a function
module.exports = resultsProcessor;
