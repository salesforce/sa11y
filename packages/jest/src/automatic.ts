/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { AxeResults, log, useCustomRules } from '@sa11y/common';
import { getViolationsJSDOM } from '@sa11y/assert';
import { A11yError, exceptionListFilterSelectorKeywords } from '@sa11y/format';
import { isTestUsingFakeTimer } from './matcher';
import { expect } from '@jest/globals';
import { adaptA11yConfig, adaptA11yConfigCustomRules } from './setup';
import { defaultRuleset } from '@sa11y/preset-rules';

/**
 * Options for Automatic checks to be passed to {@link registerSa11yAutomaticChecks}
 */
export type AutoCheckOpts = {
    runAfterEach?: boolean;
    cleanupAfterEach?: boolean;
    consolidateResults?: boolean;
    // TODO (feat): add support for optional exclusion of selected tests
    // excludeTests?: string[];
    // List of test file paths (as regex) to filter for automatic checks
    filesFilter?: string[];
    runDOMMutationObserver?: boolean;
};

/**
 * Default options when {@link registerSa11yAutomaticChecks} is invoked
 */
const defaultAutoCheckOpts: AutoCheckOpts = {
    runAfterEach: true,
    cleanupAfterEach: true,
    consolidateResults: true,
    filesFilter: [],
};

let originalDocumentBodyHtml: string | null = null;

let mutatedNodes: string[] = [];

export const setOriginalDocumentBodyHtml = (bodyHtml: string | null) => {
    originalDocumentBodyHtml = bodyHtml ?? null;
};

export const getOriginalDocumentBodyHtml = () => {
    return originalDocumentBodyHtml;
};

/**
 * Check if current test file needs to be skipped based on any provided filter
 */
export function skipTest(testPath: string | undefined, filesFilter?: string[]): boolean {
    if (!testPath || !filesFilter || !(filesFilter?.length > 0)) return false;
    const skipTest = filesFilter.some((fileName) => testPath.toLowerCase().includes(fileName.toLowerCase()));

    if (skipTest) {
        log(
            `Skipping automatic accessibility check on ${testPath} as it matches given files filter: ${filesFilter.toString()}`
        );
    }
    return skipTest;
}

/**
 * Run accessibility check on each element node in the body using {@link toBeAccessible}
 * @param opts - Options for automatic checks {@link AutoCheckOpts}
 */
export async function automaticCheck(opts: AutoCheckOpts = defaultAutoCheckOpts): Promise<void> {
    if (skipTest(expect.getState().testPath, opts.filesFilter)) return;

    // Skip automatic check if test is using fake timer as it would result in timeout
    if (isTestUsingFakeTimer()) {
        log('Skipping automatic accessibility check as Jest fake timer is in use.');
        return;
    }

    let violations: AxeResults = [];
    const currentDocumentHtml = document.body.innerHTML;
    if (originalDocumentBodyHtml) {
        document.body.innerHTML = originalDocumentBodyHtml;
    }
    // Create a DOM walker filtering only elements (skipping text, comment nodes etc)
    const walker = document.createTreeWalker(document.body, NodeFilter.SHOW_ELEMENT);
    let currNode = walker.firstChild();
    const customRules = useCustomRules();
    try {
        if (!opts.runDOMMutationObserver) {
            while (currNode !== null) {
                // TODO (spike): Use a logger lib with log levels selectable at runtime
                // console.log(
                //     `â™¿ [DEBUG] Automatically checking a11y of ${currNode.nodeName}
                //      for test "${expect.getState().currentTestName}"
                //      : ${testPath}`
                // );
                // W-10004832 - Exclude descendancy based rules from automatic checks
                if (customRules.length === 0)
                    violations.push(...(await getViolationsJSDOM(currNode, adaptA11yConfig(defaultRuleset))));
                else
                    violations.push(
                        ...(await getViolationsJSDOM(currNode, adaptA11yConfigCustomRules(defaultRuleset, customRules)))
                    );
                currNode = walker.nextSibling();
            }
        } else {
            // loop mutated nodes
            for (let i = 0; i < mutatedNodes.length; i++) {
                if (mutatedNodes[i]) {
                    document.body.innerHTML = mutatedNodes[i];
                    const walker = document.createTreeWalker(document.body, NodeFilter.SHOW_ELEMENT);
                    let currNode = walker.firstChild();
                    while (currNode !== null) {
                        if (customRules.length === 0)
                            violations.push(...(await getViolationsJSDOM(currNode, adaptA11yConfig(defaultRuleset))));
                        else
                            violations.push(
                                ...(await getViolationsJSDOM(
                                    currNode,
                                    adaptA11yConfigCustomRules(defaultRuleset, customRules)
                                ))
                            );
                        currNode = walker.nextSibling();
                    }
                }
            }
        }
    } finally {
        if (opts.runDOMMutationObserver) {
            mutatedNodes = [];
        }
        setOriginalDocumentBodyHtml(null);
        document.body.innerHTML = currentDocumentHtml;
        if (opts.cleanupAfterEach) document.body.innerHTML = ''; // remove non-element nodes
        // TODO (spike): Disable stack trace for automatic checks.
        //  Will this affect all errors globally?
        // Error.stackTraceLimit = 0;
        if (process.env.SELECTOR_FILTER_KEYWORDS) {
            violations = exceptionListFilterSelectorKeywords(
                violations,
                process.env.SELECTOR_FILTER_KEYWORDS.split(',')
            );
        }
        A11yError.checkAndThrow(violations, { deduplicate: opts.consolidateResults });
    }
}

function observerCallback(mutations: MutationRecord[], _observer: MutationObserver) {
    for (const mutation of mutations) {
        mutation.addedNodes.forEach((node) => {
            if (node?.parentElement?.innerHTML) {
                mutatedNodes.push(node.parentElement.innerHTML);
            }
        });
    }
}

// https://developer.mozilla.org/en-US/docs/Web/API/MutationObserverInit
const observerOptions: MutationObserverInit = {
    subtree: true, // extend monitoring to the entire subtree of nodes rooted at target
    childList: true, // monitor target node for addition/removal of child nodes
    attributes: true, // monitor changes to the value of attributes of nodes
    characterData: true, // monitor changes to the character data contained within nodes
};

/**
 * Register accessibility checks to be run automatically after each test
 * @param opts - Options for automatic checks {@link AutoCheckOpts}
 */
export function registerSa11yAutomaticChecks(opts: AutoCheckOpts = defaultAutoCheckOpts): void {
    if (opts.runAfterEach) {
        const observer = new MutationObserver(observerCallback);
        // TODO (fix): Make registration idempotent
        log('Registering sa11y checks to be run automatically after each test');

        beforeEach(() => {
            if (opts.runDOMMutationObserver) {
                observer.observe(document.body, observerOptions);
            }
        });

        afterEach(async () => {
            if (opts.runDOMMutationObserver) {
                observer.disconnect(); // stop mutation observer
            }
            await automaticCheck(opts);
        });
    }
}
