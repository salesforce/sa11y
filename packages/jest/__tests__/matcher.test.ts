/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { matcherHintMsg, toBeAccessible } from '../src/matcher';
import { registerA11yMatchers } from '../src';
import { extended, recommended } from '@sa11y/preset-rules';
import {
    beforeEachSetup,
    cartesianProduct,
    domWithA11yIssues,
    domWithNoA11yIssues,
    domWithA11yIssuesBodyID,
    shadowDomID,
} from '@sa11y/test-utils';

// Collection of values to be tested passed in as different API parameters
const a11yConfigParams = [extended, recommended, undefined];
const domParams = [document, undefined];
const domConfigParams = cartesianProduct(domParams, a11yConfigParams);

beforeAll(() => {
    registerA11yMatchers();
});

beforeEach(() => {
    beforeEachSetup();
});

describe('a11y matchers', () => {
    it('should be extendable with expect', () => {
        // Mostly here for code cov as it doesn't register correctly with just registerA11yMatchers()
        expect.extend({ toBeAccessible });
    });
});

describe('toBeAccessible jest a11y matcher', () => {
    it.each(domConfigParams)('should not throw error for dom with no a11y issues (arg: %#)', async (dom, config) => {
        document.body.innerHTML = domWithNoA11yIssues;
        await expect(dom).toBeAccessible(config);
    });

    it.each(a11yConfigParams)('should throw error for dom with a11y issues with config: %#', async (config) => {
        document.body.innerHTML = domWithA11yIssues;
        expect.assertions(3);
        // using the 'not' matcher just for testing, not expecting this to be used out of the unit testing context
        await expect(document).not.toBeAccessible(config);
        // using without the 'not' matcher which should be the primary way the API is used (without error catching)
        try {
            await expect(document).toBeAccessible(config);
        } catch (e) {
            expect(e.message).toContain(matcherHintMsg);
        }
    });

    it.each([
        // dom with no issues won't result in error thrown and hence will have 1 less assertion
        [shadowDomID, domWithNoA11yIssues, 2],
        [domWithA11yIssuesBodyID, domWithA11yIssues, 3],
    ])('should be able to check a11y of a HTML element: %#', async (id: string, dom: string, numAssertions: number) => {
        expect.assertions(numAssertions);
        document.body.innerHTML = dom;
        const elem = document.getElementById(id);
        expect(elem).toBeDefined();
        try {
            await expect(elem).toBeAccessible();
        } catch (e) {
            expect(e.message).toContain(matcherHintMsg);
        }
    });
});
