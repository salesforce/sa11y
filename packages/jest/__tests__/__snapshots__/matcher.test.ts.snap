// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toBeAccessible jest a11y matcher should be able to check a11y of a HTML element: dom-with-issues 1`] = `
"[2mexpect([22m[31m1 issues[39m[2m)toBeAccessible: expected document to have no accessibility violations but found 1 issues: [22m
[2m[22m
[2m [31m\\"â­• (link-name) Links must have discernible text: a\\"[39m[22m
[2m	ðŸ”— Help URL: https://dequeuniversity.com/rules/axe/3.5/link-name([22m[32m0 issues[39m[2m)[22m"
`;

exports[`toBeAccessible jest a11y matcher should throw error for dom with a11y issues with config: 0 1`] = `
"[2mexpect([22m[31m5 issues[39m[2m)toBeAccessible: expected document to have no accessibility violations but found 5 issues: [22m
[2m[22m
[2m [31m\\"â­• (bypass) Page must have means to bypass repeated blocks: html\\"[39m[22m
[2m	ðŸ”— Help URL: https://dequeuniversity.com/rules/axe/3.5/bypass[22m
[2m[22m
[2m[31m\\"â­• (document-title) Documents must have <title> element to aid in navigation: html\\"[39m[22m
[2m	ðŸ”— Help URL: https://dequeuniversity.com/rules/axe/3.5/document-title[22m
[2m[22m
[2m[31m\\"â­• (link-name) Links must have discernible text: a\\"[39m[22m
[2m	ðŸ”— Help URL: https://dequeuniversity.com/rules/axe/3.5/link-name[22m
[2m[22m
[2m[31m\\"â­• (landmark-one-main) Document must have one main landmark: html\\"[39m[22m
[2m	ðŸ”— Help URL: https://dequeuniversity.com/rules/axe/3.5/landmark-one-main[22m
[2m[22m
[2m[31m\\"â­• (page-has-heading-one) Page must contain a level-one heading: html\\"[39m[22m
[2m	ðŸ”— Help URL: https://dequeuniversity.com/rules/axe/3.5/page-has-heading-one([22m[32m0 issues[39m[2m)[22m"
`;

exports[`toBeAccessible jest a11y matcher should throw error for dom with a11y issues with config: 1 1`] = `
"[2mexpect([22m[31m4 issues[39m[2m)toBeAccessible: expected document to have no accessibility violations but found 4 issues: [22m
[2m[22m
[2m [31m\\"â­• (bypass) Page must have means to bypass repeated blocks: html\\"[39m[22m
[2m	ðŸ”— Help URL: https://dequeuniversity.com/rules/axe/3.5/bypass[22m
[2m[22m
[2m[31m\\"â­• (document-title) Documents must have <title> element to aid in navigation: html\\"[39m[22m
[2m	ðŸ”— Help URL: https://dequeuniversity.com/rules/axe/3.5/document-title[22m
[2m[22m
[2m[31m\\"â­• (link-name) Links must have discernible text: a\\"[39m[22m
[2m	ðŸ”— Help URL: https://dequeuniversity.com/rules/axe/3.5/link-name[22m
[2m[22m
[2m[31m\\"â­• (page-has-heading-one) Page must contain a level-one heading: html\\"[39m[22m
[2m	ðŸ”— Help URL: https://dequeuniversity.com/rules/axe/3.5/page-has-heading-one([22m[32m0 issues[39m[2m)[22m"
`;

exports[`toBeAccessible jest a11y matcher should throw error for dom with a11y issues with config: 2 1`] = `
"[2mexpect([22m[31m5 issues[39m[2m)toBeAccessible: expected document to have no accessibility violations but found 5 issues: [22m
[2m[22m
[2m [31m\\"â­• (bypass) Page must have means to bypass repeated blocks: html\\"[39m[22m
[2m	ðŸ”— Help URL: https://dequeuniversity.com/rules/axe/3.5/bypass[22m
[2m[22m
[2m[31m\\"â­• (document-title) Documents must have <title> element to aid in navigation: html\\"[39m[22m
[2m	ðŸ”— Help URL: https://dequeuniversity.com/rules/axe/3.5/document-title[22m
[2m[22m
[2m[31m\\"â­• (link-name) Links must have discernible text: a\\"[39m[22m
[2m	ðŸ”— Help URL: https://dequeuniversity.com/rules/axe/3.5/link-name[22m
[2m[22m
[2m[31m\\"â­• (landmark-one-main) Document must have one main landmark: html\\"[39m[22m
[2m	ðŸ”— Help URL: https://dequeuniversity.com/rules/axe/3.5/landmark-one-main[22m
[2m[22m
[2m[31m\\"â­• (page-has-heading-one) Page must contain a level-one heading: html\\"[39m[22m
[2m	ðŸ”— Help URL: https://dequeuniversity.com/rules/axe/3.5/page-has-heading-one([22m[32m0 issues[39m[2m)[22m"
`;
